runtime/src/epoch_rewards_hasher.rs

```diff
@@ -0,0 +1,266 @@
+    siphasher::sip::SipHasher13,
+    solana_sdk::{hash::Hash, pubkey::Pubkey},
+    std::hash::Hasher,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct EpochRewardHasher {
+    hasher: SipHasher13,
+    partitions: usize,
+}
+
+impl EpochRewardHasher {
+    /// Use SipHasher13 keyed on the `seed` for calculating epoch reward partition
+    pub(crate) fn new(partitions: usize, hash: &Hash) -> Self {
+        let mut hasher = SipHasher13::new();
+        hasher.write(hash.as_ref());
+        Self { hasher, partitions }
+    }
+
+    /// Return partition index (0..partitions) by hashing `address` with the `hasher`
+    pub(crate) fn hash_address_to_partition(self, address: &Pubkey) -> usize {
+        let Self {
+            mut hasher,
+            partitions,
+        } = self;
+        hasher.write(address.as_ref());
+        let hash64 = hasher.finish();
+
+        hash_to_partition(hash64, partitions)
+    }
+}
+
+/// Compute the partition index by modulo the address hash to number of partitions w.o bias.
+/// (rand_int * DESIRED_RANGE_MAX) / (RAND_MAX + 1)
+fn hash_to_partition(hash: u64, partitions: usize) -> usize {
+    ((partitions as u128)
+        .saturating_mul(u128::from(hash))
+        .saturating_div(u128::from(u64::MAX).saturating_add(1))) as usize
+}
+
+#[allow(dead_code)]
+fn hash_rewards_into_partitions(
+    stake_rewards: &StakeRewards,
+    parent_block_hash: &Hash,
+    num_partitions: usize,
+) -> Vec<StakeRewards> {
+    let hasher = EpochRewardHasher::new(num_partitions, parent_block_hash);
+    let mut result = vec![vec![]; num_partitions];
+
+    for reward in stake_rewards {
+        let partition_index = hasher.hash_address_to_partition(&reward.stake_pubkey);
+        result[partition_index].push(reward.clone());
+    }
+    result
+}
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*,
+        crate::stake_rewards::StakeReward,
+        std::{collections::HashMap, ops::RangeInclusive},
+    };
+
+    #[test]
+    fn test_get_equal_partition_range() {
+        // show how 2 equal partition ranges are 0..=(max/2), (max/2+1)..=max
+        // the inclusive is tricky to think about
+        let range = get_equal_partition_range(0, 2);
+        assert_eq!(*range.start(), 0);
+        assert_eq!(*range.end(), u64::MAX / 2);
+        let range = get_equal_partition_range(1, 2);
+        assert_eq!(*range.start(), u64::MAX / 2 + 1);
+        assert_eq!(*range.end(), u64::MAX);
+    }
+
+    #[test]
+    fn test_hash_to_partitions() {
+        let partitions = 16;
+        assert_eq!(hash_to_partition(0, partitions), 0);
+        assert_eq!(hash_to_partition(u64::MAX / 16, partitions), 0);
+        assert_eq!(hash_to_partition(u64::MAX / 16 + 1, partitions), 1);
+        assert_eq!(hash_to_partition(u64::MAX / 16 * 2, partitions), 1);
+        assert_eq!(hash_to_partition(u64::MAX / 16 * 2 + 1, partitions), 1);
+        assert_eq!(hash_to_partition(u64::MAX - 1, partitions), partitions - 1);
```

Comment:
these aren't really the boundaries i'd expect. maybe behzad was right about the math seeming off :thinking:

Reply:
I put up more tests verifying the regions to be of the size I expected and matching what I expected.
What were you expecting?

Reply:
n/m i just wasn't accounting for `RAND_MAX % PARTITIONS` skewing some of the partions' size by one